dosbox-staging/dosbox-staging
# Move environment functions to psp (+ other small changes)

# Description

Most of the changes involve moving the functions querying and commanding the environment out of the shell class and into the psp class. These function only operate on the shell's psp anyways so it makes no sense to have them outside the psp. This also means that programs that query the environment no longer have to go through the hacky global `first_shell`. Removing `first_shell` was the primary motivation for these commits, though after these changes it is still being used in autoexec.

There are also a few single-commit changes in here:

Moved shell_history out of individual functions and into its own class.

Fixed a crash in the setver program when too few args are present.

Vastly improved the semantics of the FileReader class (the interface remains unchanged).

More details in a comment below.

# Manual testing

Ensure that read, write, and list environment functions work as expected with the `SET` command.

- `SET PATH`
- `SET VAR=VAL`
- `SET`

`MOUNT -Z` changes the letter of the Z drive. Note that you can't run this more than once per session. From what I can tell, this was a hack for WINE, since there was a conflict with the WINE Z drive. In addition to the environment interface change, I made the change propagate across subshells, whereas it would break your path before.

 - Run `MOUNT -Z E`. Ensure that your PATH variable has changed to include E: instead of Z:. Go into a subshell and ensure that the change is visible there too.
 - Restart and run the command again, this time from a subshell. Ensure the changes are reflected in the parent shell.

ParseLine is functionally unchanged. Only the name resolution for temp files was touched. Launch from terminal so you can see logging ouptut. Input any command with the vertical bar pipe. It can be a garbage command like `a | a`. In the logging output, you should see `Redirecting output to pipeXXXX.tmp`. Now try again while setting a `TMP` or `TEMP` environment variable to a directory. If you pipe again, the file will be prefixed by the chosen directory.

Environment variable resolution is functionally unchanged. Create a batch file with the line `ECHO %PATH%`, and ensure that your path is printed. Do the same directly on the DOS command line when the settings to get shell expansion there are set to on.

Executable resolution is functionally unchanged. Ensure all directories in your `PATH` work.

The `PATH` command is functionally unchanged. Ensure it matches the output of `SET PATH`.

The `DIR` command is functionally unchanged. Ensure that it works without the `DIRCMD` environment variable being set. Also ensure that it applies the expected switches when `DIRCMD` is set. For example, `SET DIRCMD=/b` should apply the `/b` switch automatically when using `DIR`.

The `CONFIG` command is functionally unchanged. The writes to the `CONFIG` environment variable should be the same as before.

- `CONFIG -GET VER` echoes current does version, and sets the `CONFIG` environment variable to that value
- `CONFIG -GET DOS VER` does the same

## Shell History Tests
Make sure the config file can be read from and written to.

When the sum of lines in the history file and the new commands are over 500 (or whatever the specified constant in the class is), ensure that the file contains only the most recent 500 lines upon exit.

Ensure that exit never written to the file.

Remove read permissions and ensure that dosbox doesn't explode. Then do that again, but make the file write-only.

# Checklist

I have:
